pennylane/labs/vibrational/christiansen_utils.py

```diff
@@ -0,0 +1,911 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utility functions related to the construction of the taylor form Hamiltonian."""
+import numpy as np
+import h5py
+from scipy.special import factorial
+import itertools
+import subprocess
+from mpi4py import MPI
+
+comm = MPI.COMM_WORLD
+rank = comm.Get_rank()
+size = comm.Get_size()
+
+
+def _cform_onemode_kinetic(freqs, n_states):
+    """Calculates the kinetic energy part of the one body integrals to correct the integrals
+    for localized modes
+
+    Args:
+        freqs(int): the harmonic frequencies
+        n_states(int): maximum number of bosonic states per mode
+
+    Returns:
+        the kinetic energy part of the one body integrals
+    """
+    # action of kinetic energy operator for m=1,...,M modes with frequencies freqs[m]
+    nmodes = len(freqs)
+    all_mode_combos = []
+    for aa in range(nmodes):
+        all_mode_combos.append([aa])
+    all_bos_combos = list(itertools.product(range(n_states), range(n_states)))
+
+    rank = comm.Get_rank()
+    size = comm.Get_size()
+    boscombos_on_rank = np.array_split(all_bos_combos, size)[rank]
+    chunksize = len(boscombos_on_rank)
+
+    local_K_mat = np.zeros(len(all_mode_combos) * chunksize)
+    for nn, [ii] in enumerate(all_mode_combos):
+        ii = int(ii)
+        m_const = freqs[ii] / 4
+
+        for mm, [ki, hi] in enumerate(boscombos_on_rank):
+            ind = nn * len(boscombos_on_rank) + mm
+
+            if ki == hi:
+                local_K_mat[ind] += m_const * (2 * ki + 1)
+            elif ki == hi + 2:
+                local_K_mat[ind] -= m_const * np.sqrt((hi + 2) * (hi + 1))
+            elif ki == hi - 2:
+                local_K_mat[ind] -= m_const * np.sqrt((hi - 1) * hi)
+    return local_K_mat
+
+
+def _cform_twomode_kinetic(pes, n_states):
+    """Calculates the kinetic energy part of the two body integrals to correct the integrals
+    for localized modes
+
+    Args:
+        pes(VibrationalPES): object containing the vibrational potential energy surface data
+        n_states(int): maximum number of bosonic states per mode
+
+    Returns:
+        the kinetic energy part of the two body integrals
+    """
+    nmodes = len(pes.freqs)
+
+    all_mode_combos = []
+    for aa in range(nmodes):
+        for bb in range(nmodes):
+            all_mode_combos.append([aa, bb])
+    all_bos_combos = list(
+        itertools.product(range(n_states), range(n_states), range(n_states), range(n_states))
+    )
+
+    rank = comm.Get_rank()
+    size = comm.Get_size()
+    boscombos_on_rank = np.array_split(all_bos_combos, size)[rank]
+    chunksize = len(boscombos_on_rank)
+
+    local_kin_cform_twobody = np.zeros(len(all_mode_combos) * chunksize)
+
+    for nn, (ii, jj) in enumerate(all_mode_combos):
+        ii, jj = int(ii), int(jj)
+
+        # Skip unnecessary combinations
+        if jj >= ii:
+            continue
```

Comment:
I think we can skip generating these unnecessary combinations, will improve the scaling for bigger systems, maybe something to take care of when we move things to pennylane. We can add a todo here for now.

Reply:
I made this change in the PES functions, we can follow the same process.